loki/tests/test_subroutine.py

```diff
@@ -2238,3 +2239,75 @@ def test_call_args_kwargs_conversion(frontend):
     for call in FindNodes(ir.CallStatement).visit(driver.body):
         assert tuple(arg.name for arg in call.arguments) == call_args
         assert call.kwarguments == ()
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_resolve_typebound_var(frontend, tmp_path):
+    """
+    Test correct behaviour of :any:`Scope.resolve_typebound_var` utility
+    """
+    fcode = """
+module header_mod
+    implicit none
+    type some_type
+        integer :: ival
+    end type some_type
+
+    type other_type
+        type(some_type) :: other
+    end type other_type
+
+    type third_type
+        type(other_type) :: some
+    end type third_type
+end module header_mod
+
+subroutine some_routine
+    use header_mod, only: third_type
+    implicit none
+    type(third_type) :: tt
+end subroutine
+    """.strip()
+
+    source = Sourcefile.from_source(fcode, frontend=frontend, xmods=[tmp_path])
+    # routine = Subroutine.from_source(fcode_routine, frontend=frontend, xmods=[tmp_path], definitions=source.definitions) #source['some_routine']
+    routine = source['some_routine']
+
+    tt_some = routine.resolve_typebound_var('tt%some')
+    assert tt_some == 'tt%some'
+    assert tt_some.type.dtype.name == 'other_type'
+    assert tt_some.type.dtype.typedef is source['header_mod']['other_type']
+
+    tt_some_other_ival = routine.resolve_typebound_var('tt%some%other%ival')
+    assert tt_some_other_ival == 'tt%some%other%ival'
+    assert tt_some_other_ival.type.dtype == BasicType.INTEGER
+    assert tt_some_other_ival.parent.type.dtype.name == 'some_type'
+    assert tt_some_other_ival.parent.type.dtype.typedef is source['header_mod']['some_type']
+
+    tt = routine.resolve_typebound_var('tt')
+    assert tt == 'tt'
+    assert tt.type.dtype.name == 'third_type'
+    assert tt.type.dtype.typedef is source['header_mod']['third_type']
+
+    # This does not throw an error as the use-case of incomplete type definitions
+    # may well require working with incomplete type definitions
+    tt_invalid_val = routine.resolve_typebound_var('tt%invalid%val')
+    assert tt_invalid_val == 'tt%invalid%val'
```

Comment:
I'm not sure I understand the reasoning here. The type of `tt` is known so it can be safely inferred that `tt%invalid` is in deed invalid. If the type of `tt` were deferred, I'd agree, but it is attached in this case, no?

Reply:
Right, I see your point. I was thinking of situations where we don't have the type definition and overlooked the fact that we actually have the complete definition here. Let me see if I can quickly add an error in this case without harming the situation where we don't have the full type definition.

Reply:
Looks great. Many thanks for addressing this.